File: mobile/lib/protocol/welcome_payload.dart
import 'dart:typed_data';

import 'frame.dart';
import 'payload_codec.dart';

class WelcomePayload {
  const WelcomePayload({
    required this.serverName,
    required this.effectiveMajor,
    required this.effectiveMinor,
    required this.minSupportedMajor,
    required this.sessionId,
    required this.authRequired,
    required this.challenge,
  });

  final String serverName;
  final int effectiveMajor;
  final int effectiveMinor;
  final int minSupportedMajor;
  final Uint8List sessionId;
  final bool authRequired;
  final Uint8List challenge;

  @override
  bool operator ==(Object other) =>
      other is WelcomePayload &&
      serverName == other.serverName &&
      effectiveMajor == other.effectiveMajor &&
      effectiveMinor == other.effectiveMinor &&
      minSupportedMajor == other.minSupportedMajor &&
      _bytesEqual(sessionId, other.sessionId) &&
      authRequired == other.authRequired &&
      _bytesEqual(challenge, other.challenge);

  @override
  int get hashCode => Object.hash(
        serverName,
        effectiveMajor,
        effectiveMinor,
        minSupportedMajor,
        Object.hashAll(sessionId),
        authRequired,
        Object.hashAll(challenge),
      );
}

class WelcomePayloadCodec {
  static Uint8List encode(WelcomePayload payload) {
    if (payload.sessionId.length != 16) {
      throw const ProtocolException('sessionId must be exactly 16 bytes.');
    }
    if (payload.challenge.length != 32) {
      throw const ProtocolException('challenge must be exactly 32 bytes.');
    }

    final writer = PayloadWriter()
      ..writeString(payload.serverName, minLength: 1, maxLength: 64)
      ..writeUInt8(payload.effectiveMajor)
      ..writeUInt8(payload.effectiveMinor)
      ..writeUInt8(payload.minSupportedMajor)
      ..writeBytes(payload.sessionId)
      ..writeUInt8(payload.authRequired ? 1 : 0)
      ..writeBytes(payload.challenge);
    return writer.toBytes();
  }

  static WelcomePayload decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final serverName = reader.readString(minLength: 1, maxLength: 64);
    final effectiveMajor = reader.readUInt8();
    final effectiveMinor = reader.readUInt8();
    final minSupportedMajor = reader.readUInt8();
    final sessionId = reader.readBytes(16);
    final authRequired = reader.readUInt8();
    final challenge = reader.readBytes(32);
    reader.expectEnd();

    if (authRequired > 1) {
      throw const ProtocolException('authRequired must be 0 or 1.');
    }

    return WelcomePayload(
      serverName: serverName,
      effectiveMajor: effectiveMajor,
      effectiveMinor: effectiveMinor,
      minSupportedMajor: minSupportedMajor,
      sessionId: sessionId,
      authRequired: authRequired == 1,
      challenge: challenge,
    );
  }
}

bool _bytesEqual(Uint8List a, Uint8List b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}

File: desktop/Protocol/WelcomePayload.cs
namespace CTRL.Desktop.Protocol;

public sealed record WelcomePayload(
    string ServerName,
    byte EffectiveMajor,
    byte EffectiveMinor,
    byte MinSupportedMajor,
    byte[] SessionId,
    bool AuthRequired,
    byte[] Challenge);

public static class WelcomePayloadCodec
{
    public static byte[] Encode(WelcomePayload payload)
    {
        if (payload.SessionId is null || payload.SessionId.Length != 16)
            throw new ProtocolException("sessionId must be exactly 16 bytes.");
        if (payload.Challenge is null || payload.Challenge.Length != 32)
            throw new ProtocolException("challenge must be exactly 32 bytes.");

        var writer = new PayloadWriter();
        writer.WriteString(payload.ServerName, 1, 64);
        writer.WriteUInt8(payload.EffectiveMajor);
        writer.WriteUInt8(payload.EffectiveMinor);
        writer.WriteUInt8(payload.MinSupportedMajor);
        writer.WriteBytes(payload.SessionId);
        writer.WriteUInt8(payload.AuthRequired ? (byte)1 : (byte)0);
        writer.WriteBytes(payload.Challenge);
        return writer.ToArray();
    }

    public static WelcomePayload Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var serverName = reader.ReadString(1, 64);
        var effectiveMajor = reader.ReadUInt8();
        var effectiveMinor = reader.ReadUInt8();
        var minSupportedMajor = reader.ReadUInt8();
        var sessionId = reader.ReadBytes(16);
        var authRequired = reader.ReadUInt8();
        var challenge = reader.ReadBytes(32);
        reader.ExpectEnd();

        if (authRequired > 1)
            throw new ProtocolException("authRequired must be 0 or 1.");

        return new WelcomePayload(
            serverName,
            effectiveMajor,
            effectiveMinor,
            minSupportedMajor,
            sessionId,
            authRequired == 1,
            challenge);
    }
}
